### PR TITLE
Add LASSO-regularized propensity score (g-model) strategy (LassoCTMLE) 

### PR DESCRIPTION
**Pull request by Asantewaah**
## Summary

This draft PR introduces the collaborative strategy, `LassoCTMLE`, to the TMLE.jl package.  
`LassoCTMLE` fits a LASSO-regularized logistic regression model (using GLMNet.jl) for the propensity score (g-model), with cross-validated lambda selection. The strategy is intended to be generic, supporting any parameter Ψ (not just ATE), and integrates with the existing TMLE.jl API, as advised.

## Motivation for this estimator

- Enables flexible, regularized estimation of the g-model, which can improve robustness in high-dimensional or collinear settings.
- Helps users avoid overfitting and select an appropriate level of regularization via cross-validation.

## Implementation Details

- It uses cross-validation for lambda selection.

## Testing

- I would appreciate guidance on best practices for testing strategies in TMLE.jl.
- Are there recommended datasets or paradigms in the `test/` folder I should follow?
- What are the edge cases or parameterizations I should cover?
- Should I benchmark against existing strategies, and if so, with which datasets?

## Open Questions

- I currently get `UndefVarError: TMLE not defined in Main` when I try to run the singular lasso_strategy script, just to check if everything is imported or written well. What's the natural way to import or use the package to test the code I write?
- I'm not sure if I succeeded in making the current constructor/API for `LassoCTMLE` idiomatic for TMLE.jl?

## Next Steps

- Integrate feedback from Olivier.
- Add more tests and documentation as needed.

Thank you for reviewing.

### CODE REVIEW

**Comment by Asantewaah** [comment]
> No problem! Do you have an test idea? That is, a simulation setting in which you expect this Lasso-CTMLE to perform better than the canonical one?

Yes, I actually have a setup I have been using ever since I started building the estimator; it's basically a simulation using the Toeplitz function so we can adjust or vary the correlation parameter. From my tests, the higher the correlation, rho, between the selection covariates, the better C-TMLE performs compared to TMLE. I will add the Julia converted code to the test/lasso_strategy.jl. Still going through how to actually test anything new I add to the package.

**Comment by olivierlabayle** [comment]
In order. to test you simply use the estimator on your generated dataset, for the structure of the file, you can basically look at any other test file, for instance this is an easy one to read: https://github.com/TARGENE/TMLE.jl/blob/main/test/counterfactual_mean_based/non_regression_test.jl

**Comment by olivierlabayle** [comment]
> Hi @olivierlabayle ,
> 
> I tried to mirror the syntax used in `non_regression_test.jl` for my LassoCTMLE test, as suggested. However, I keep getting this recurring error:
> 
> ```
> ERROR: UndefVarError: `LassoCTMLE` not defined in `Main`
> ```
> 
> I also tried calling the other strategies you created for the template much earlier (`AdaptiveCorrelationStrategy`, `GreedyStrategy`), and they’re giving the same error, even though they are exported and their files are included in `TMLE.jl`.
> 
> Do you have any idea why these strategies aren’t visible, or what I might be missing?
> 
> Thanks

You need to push a test file so that I can see what you are doing, from your branch, this works perfectly fine:

image

The reason why your strategy is not visible is because you haven't included the source file in the package. If you include the file you will likely have an error when trying to import the package because you import other packages that are not in the Projrct.toml. As a reminder, I suggested reading the contribution guide here: https://targene.github.io/TMLE.jl/previews/PR135/contributing/. I linked a youtube video which is worth watching to understand the basics of Julia packages.

**Comment by Asantewaah** [comment]
> > Hi @olivierlabayle ,
> > I tried to mirror the syntax used in `non_regression_test.jl` for my LassoCTMLE test, as suggested. However, I keep getting this recurring error:
> > ```
> > ERROR: UndefVarError: `LassoCTMLE` not defined in `Main`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I also tried calling the other strategies you created for the template much earlier (`AdaptiveCorrelationStrategy`, `GreedyStrategy`), and they’re giving the same error, even though they are exported and their files are included in `TMLE.jl`.
> > Do you have any idea why these strategies aren’t visible, or what I might be missing?
> > Thanks
> 
> You need to push a test file so that I can see what you are doing, from your branch, this works perfectly fine:
> 
> image
> The reason why your strategy is not visible is because you haven't included the source file in the package. If you include the file you will likely have an error when trying to import the package because you import other packages that are not in the Projrct.toml. As a reminder, I suggested reading the contribution guide here: https://targene.github.io/TMLE.jl/previews/PR135/contributing/. I linked a youtube video which is worth watching to understand the basics of Julia packages.


I did watch the video, really detailed but I don't understand why (`AdaptiveCorrelationStrategy`,and `GreedyStrategy`) is not registed when I call them even though I did add those. before I push the test files, I try to run it locally first so I don't mess up anything. I just commited it.

**Comment by Asantewaah** [comment]
Hi @olivierlabayle ,

I’ve resolved my environment issues and some other syntax inconsistencies with my code, but now the LassoCTMLE test fails with:

```
MethodError: no method matching glmnet!(::Matrix{Float64}, ::Vector{Float64}, ::Binomial{Float64}; lambda::Vector{Float64})
```

I’ve tried both `glmnet` and `glmnet!` but still get the same error about argument types.  
Pushing my code now for you to see.

Thanks.

File: src/TMLE.jl [hunk 1]
```diff
@@ -23,6 +23,8 @@ using OrderedCollections
 using AutoHashEquals
 using StatisticalMeasures
 using DataFrames
+using GLMNet
```
[the remark this answers is not in the document]
Comment (olivierlabayle) [reply src/TMLE.jl:26]: Also just realising this issue might be due to type requirements, you will need to provide the full error message for me to be able to help more here

File: src/TMLE.jl [hunk 2]
```diff
@@ -23,6 +23,8 @@ using OrderedCollections
...
+using ToeplitzMatrices
```
[the remark this answers is not in the document]
Comment (Asantewaah) [reply src/TMLE.jl:27]: I use the ToeplitzMatrices to generate a correlation matrix for the covariates and be able to control the correlation between the covariates

File: Project.toml [hunk 3]
```diff
@@ -28,6 +29,7 @@ StatisticalMeasures = "a19d573c-0a75-4610-95b3-7071388c7541"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 TableOperations = "ab02a1b2-a7df-11e8-156e-fb1833f50b87"
 Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
+ToeplitzMatrices = "c751599d-da0a-543b-9d20-d0a503d91d24"
```
Comment (olivierlabayle) [inline Project.toml:32]: As discussed if really needed, this should be in `test/Project.toml`. To add a test dependency you can do:

```julia
]activate test
add ToeplitzMatrices
```

Note that if you want to generate correlated treatments, you don't necessarily need an external package, see for example: https://github.com/TARGENE/TMLE.jl/blob/aeab2c16b0ab3970ac46bd719c823573ea620ac8/examples/interactions_correlated.jl#L89

File: test/counterfactual_mean_based/lasso_strategy.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        treatment_confounders = (A = confounders,)
+    )
+    lasso_estimator = Tmle(
+        collaborative_strategy = LassoCTMLE(
+            confounders = confounders,
+            dataset = dataset
+        )
+    )
+    lasso_result, _ = lasso_estimator(Ψ, dataset; verbosity = 0)
+    @test !isnan(estimate(lasso_result))
```
Comment (olivierlabayle) [inline test/counterfactual_mean_based/lasso_strategy.jl:55]: Can you notice any improvement using this estimator e.g., coverage ? bias ? variance ?

**Review by olivierlabayle — COMMENTED** [review]
Thanks Asantewaa, looks like you are in a good position to make some progress on testing your implementation now!

File: test/counterfactual_mean_based/lasso_strategy.jl (hunk 4 above)
Comment (Asantewaah) [reply test/counterfactual_mean_based/lasso_strategy.jl:54]: I still haven't passed the test; there seems to be an issue with how the data is being passed to the propensity score. I will take the package (Toeplitz) out of the main environment and add it to the test environment instead.

Comment (Asantewaah) [reply test/counterfactual_mean_based/lasso_strategy.jl:54]: Hi @olivierlabayle ,

I’ve been away from research the past couple of weeks attending trainings and conferences, but I’m back now. I’ve added the LinearAlgebra package, but I’m now getting a “Not Implemented Error”. Could you please help me with this when you have a chance?

**Comment by Asantewaah** [comment]
Hi @olivierlabayle ,

I think the integration is working fine now. The code is a little messy, but it runs and all the tests pass. I also added an example in the `examples` folder, where I created a function for the Toeplitz simulation and compared its estimates to TMLE. The only thing I ran into was with the lambda values for the lasso regression, the automatically(using cross validation) generated ones weren’t being applied properly during the updating step, which made the results come out the same as TMLE. To work around this, I set up a way to manually specify lambda values, and for now there’s a mix of manual and automatic comparison logic (you’ll see that in the tests).

If you could take a look when you have a chance, that’d be great.

**Comment by Asantewaah** [comment]
Hi @olivierlabayle 

So I've got a bootstrap analysis working with some nice visualizations. The results look good.

I did ran into was some dependency conflicts with ToeplitzMatrices.jl - it kept causing issues, so I just wrote a simple `create_toeplitz_matrix()` function from scratch instead. 

I am getting some package extension warnings during the environment switching (stuff like IntervalSets, Zygote extensions complaining), but everything actually works.

I also made sure the verbose control system works well, so by default everything runs silently for production use, but you can flip `verbose=true` if you want to see all the variable selection details during development(the generated lambda, the glmnet results, the selected confounders......).

---

File: test/runtests.jl [hunk 5]
```diff
@@ -1,38 +1,38 @@
 using Test
 using TMLE
 
-TEST_DIR = joinpath(pkgdir(TMLE), "test")
```
Comment (olivierlabayle) [inline test/runtests.jl:4]: please revert the changes to this file. The `joinpath` is required to make sure the tests runs regardless of the run directory.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+# Example
+```julia
+strategy = LassoCTMLE(confounders = [:W1, :W2, :W3], patience = 5)
+estimator = Tmle(collaborative_strategy = strategy)
+result, _ = estimator(estimand, data)
+```
+"""
+mutable struct LassoCTMLE <: CollaborativeStrategy
+    confounders::Vector{Symbol}
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl]: why does the strategy require confounders? Should these not simply be the estimand's confounders?

**Review by olivierlabayle — CHANGES_REQUESTED** [review]
Hi Asantewaa,

Thanks for meeting today and good effort with the PR! I've left a few comments on the code as discussed today. Here are the more general notes I took as I was reviewing for completeness.

The R package for reference: https://github.com/jucheng1992/ctmle/tree/master

**Implementation**

1. Initial estimator: 
    1. Currently this is done with the user provided G model which could be anything —> problem, the lasso CTMLE must work only with a G model that is a glmnet (see point below) 
    2. I think the initial G should be fitted with glmnetcv —> need to create a GLMNet MLJ compliant model, take inspiration / use the one in TMLECLI: https://github.com/TARGENE/TMLECLI.jl/blob/main/src/models/glmnet.jl 
2. Create a sequence of estimators: the sequence of lambdas in the previous step defines the sequence
3. Evaluate on cv:
    1. I believe StepKPropensityScoreIterator should iterate only once and return a “fixed estimator” that returns the precomputed linear model from glmnetcv. The propensity score is the original propensity score.
    2. Exhausted returns true at the end of the sequence or if patience is reached
    3. This yields a lambda CTMLE and associated GLM / Qstar
4. There seems to be a termination step which I can’t see in the current implementation. I am also not entirely clear what is the logic I’ve seen in the paper.

**Tests**

- Tests are too sparse, we need to test intermediate functions and logic
- How does it work with categorical treatments? What about multiple treatments?

**Example/Docs** 

- The current comparison is with a standard estimator only using linear models (no cross validation). In order to disentangle the effect of model specification from the effect of C-TMLE, you need to provide a GLMNet to the standard estimator’s G model.
- Only code and documentation goes into the repo, hence temporary plots you make locally should not be committed to the repo. If you want to share some results with me you can add them to the PR or send them over Teams.
- In order to later integrate within the docs you don’t need to do `using Pkg` and `Pkg.activate`. The docs environment is used in the docs and the TMLE packaged is added to it with `dev`. Look at the other examples for how to make your own.

File: examples/lasso_example_old.jl
Comment (olivierlabayle) [inline examples/lasso_example_old.jl:1]: if this is old, it probably needs to be removed from the repo

File: src/TMLE.jl [hunk 7]
```diff
@@ -23,6 +23,7 @@ using OrderedCollections
 using AutoHashEquals
 using StatisticalMeasures
 using DataFrames
+import GLMNet
```
Comment (olivierlabayle) [inline src/TMLE.jl:26]: We will eventually need to make GLMNet a package extension. That is we only want to load the LassoCTMLE code when the use loads GLMNet and not have GLMnet as a direct TMLE.jl dependency. Do you think you could do that?

Some docs that can help:

- https://pkgdocs.julialang.org/v1/creating-packages/#Conditional-loading-of-code-in-packages-(Extensions)
- https://www.youtube.com/watch?v=TiIZlQhFzyk

File: examples/lasso_example.jl
Comment (olivierlabayle) [inline examples/lasso_example.jl:1]: Could you make this example compatible with and included in the docs? This is [where](https://github.com/TARGENE/TMLE.jl/blob/0e4f3cc40c270d537a2b84d781dd65979d5fed4e/docs/make.jl#L43) you can add an example file to the docs. The file must respect the [Literate.jl](https://github.com/fredrikekre/Literate.jl) format, which is a plain script. Use comments to drive the narative of what the example is and what it shows.

Comment (olivierlabayle) [reply examples/lasso_example.jl:1]: You are welcome to use the other examples as a source of inspiration to build yours.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+estimator = Tmle(collaborative_strategy = strategy)
+result, _ = estimator(estimand, data)
+```
+"""
+mutable struct LassoCTMLE <: CollaborativeStrategy
+    patience::Int
+    lambda_path::Vector{Float64}
+    cv_folds::Int
+    alpha::Float64
+    verbose::Bool
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl]: As discussed last time the verbosity level is not decided by the LassoCTMLE.

File: test/Project.toml [hunk 9]
```diff
@@ -7,6 +7,7 @@ DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
 Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
 HypothesisTests = "09f84164-cd44-5f33-b23f-e6b0d136a0d5"
 JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
```
Comment (olivierlabayle) [inline test/Project.toml]: I don't think you are using ToeplitzMatrix andLinear Algebra anymore in yours tests, so they should be removed.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        log_info(strategy, "LassoCTMLE: Checking exhaustion - iteration $(strategy.current_iteration)/$(strategy.patience), exhausted: $is_exhausted")
+    end
+    
+    return is_exhausted
+end
+
+"""
+Create propensity score specification using the given confounders list.
+"""
+function propensity_score(Ψ, confounders_list::Vector{Symbol}, strategy::LassoCTMLE)
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl:102]: I thought we agreed that you could use the default for this since the covariates list is dealt with by the lassoed model (coeff = 0)

Comment (olivierlabayle) [reply src/counterfactual_mean_based/lasso_strategy.jl:102]: No extraction of covariates is needed to fit the propensity score, the lambda will automatically set the coefficients to 0.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+end
+
+"""
+Iterator implementation for LASSO-based collaborative TMLE.
+Explores different lambda values with GLMNet regularization.
+"""
+function Base.iterate(it::TMLE.StepKPropensityScoreIterator{LassoCTMLE})
+    strategy = it.collaborative_strategy
+    
+    if isempty(strategy.lambda_path)
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl]: As discussed last time I don't think you need to refit individual lasso models if you use the path and `betas` produced by the `glmnetcv` method which is obtained from the initial fit. I think you can trick the current procedure by returning an estimators that will produce an glm model with weights given by the new lambda, regardless of the data.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    alpha::Float64
+    verbose::Bool
+    current_iteration::Int
+    explored_lambdas::Set{Float64}
+    best_lambda::Union{Float64, Nothing}
+    best_cv_loss::Float64
+
+    function LassoCTMLE(; 
+        patience = 5,
+        lambda_path = :cv,
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl]: I think these are defined by the G `glmnetcv` procedure.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        n_selected = max(1, round(Int, length(var_names) * selection_fraction))
+        return var_names[1:min(n_selected, length(var_names))], nothing
+    end
+end
+
+"""
+Extract a vector of confounder symbols from the estimand `Ψ`.
+Collects treatment-specific confounders (in order) and returns unique symbols.
+"""
+function extract_confounders_from_estimand(Ψ)
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl:74]: This is likely not needed

File: test/counterfactual_mean_based/lasso_strategy.jl
Comment (olivierlabayle) [inline test/counterfactual_mean_based/lasso_strategy.jl:1]: Overall, we will need more tests to convince ourselves and others that the code is doing what it is intended to do. As a guiding principle, think that almost **all** the functions you write should be tested if possible. For the statistical validity of the estimator, the documented example you have can suffice as it can be considered an end to end test.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        selected_indices = findall(x -> abs(x) > 1e-6, coeffs)
+        
+        if isempty(selected_indices)
+            strategy.verbose && @warn "No variables selected by GLMNet, falling back to correlation"
+            n_selected = max(1, round(Int, length(var_names) * 0.5))
+            return var_names[1:n_selected], fit
+        end
+        
+        selected_vars = var_names[selected_indices]
+        log_info(strategy, "GLMNet: α=$alpha, λ=$lambda → $(length(selected_vars))/$(length(var_names)) variables selected")
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl]: There is excessible logging in this PR which makes it difficult to navigate. Print statements that you use for your own debugging should be removed when you comit to keep the code as simple as possible.

File: src/counterfactual_mean_based/lasso_strategy.jl [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    all = Symbol[]
+    for T in Ψtreatments
+        if hasproperty(Ψ, :treatment_confounders) && haskey(Ψ.treatment_confounders, T)
+            append!(all, collect(Ψ.treatment_confounders[T]))
+        end
+    end
+    return unique(all)
+end
+
+function initialise!(strategy::LassoCTMLE, Ψ)
```
Comment (olivierlabayle) [inline src/counterfactual_mean_based/lasso_strategy.jl:85]: half of this function definition are logs. Remove.

**Review by olivierlabayle — CHANGES_REQUESTED** [review]

File: src/TMLE.jl (hunk 7 above)
Comment (Asantewaah) [reply src/TMLE.jl:26]: I can give it a try after dealing with the other comments and revisions.

**Comment by Asantewaah** [comment]
Hi @olivierlabayle , 


I've implemented your suggestion to avoid refitting, using GLMNet CV's optimal lambda directly without iterating through multiple candidates.

I had to add two small helpers (`PrefitGLMNetConditionalDistribution` in estimates.jl and `PrefitGLMNetJointConditionalDistributionEstimator` in estimators.jl) to store and reuse the CV coefficients without refitting, I hope that's alright?

Oh and I've cleaned up the `LassoCTMLE` struct by removing all the collaborative iteration stuff (`patience`, `lambda_path`, etc).

The tests all pass. 

Still working on adapting the example file for the docs and I'm yet to look into making GLMNet a package extension, it's a lot to go through but I created the GLMNetExt.jl file[it's empty] just to start
